README
Geo-Hash version 0.03

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

Alternatively, to install with Module::Build, you can use the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

DEPENDENCIES

None.

COPYRIGHT AND LICENCE

Copyright (C) 2008, Andy Armstrong

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Geo-Hash

0.01    Fri Feb 29 01:11:07 2008
        - Initial release.



php/test.php
<?php
/**
 * Geohash generation class test harness
 * http://blog.dixo.net/downloads/
 *
 * This file copyright (C) 2008 Paul Dixon (paul@elphin.com)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */


require_once('geohash.class.php');

$geohash=new Geohash;

//these test hashes were made on geohash.org
//and test various combinations of precision
//and range
$tests=array(
	"ezs42"=>array(42.6,-5.6),
	"mh7w"=>array(-20, 50),
	"t3b9m"=>array(10.1, 57.2),
	"c2b25ps"=>array(49.26, -123.26),
	"80021bgm"=>array(0.005, -179.567),
	"k484ht99h2"=>array(-30.55555, 0.2),
	"8buh2w4pnt"=>array(5.00001, -140.6),
);


foreach($tests as $actualhash=>$coords)
{
	$computed_hash=$geohash->encode($coords[0], $coords[1]);
	
	echo "Encode {$coords[0]}, {$coords[1]} as $actualhash : ";
	if ($computed_hash==$actualhash)
	{
		echo "OK<br>";
	}
	else
	{
		echo "FAIL (got $computed_hash)<br>";
	}
	
	
	echo "<hr>";
	
	$computed_coords=$geohash->decode($actualhash);
	
	echo "Decode $actualhash as {$coords[0]}, {$coords[1]} : ";
	if (($computed_coords[0]==$coords[0]) && ($computed_coords[1]==$coords[1]))
	{
		echo "OK<br>";
	}
	else
	{
		echo "FAIL (got {$computed_coords[0]}, {$computed_coords[1]})<br>";
	}

	echo "<hr>";
}



?>

andy/poly.pl
#!/usr/bin/perl

use strict;
use warnings;

$| = 1;

sub inside {
    my ($x, $y, $poly) = @_;
    
}

 FUNCTION Inside, x, y, px, py

   ;  x - The x coordinate of the point.
   ;  y - The y coordinate of the point.
   ; px - The x coordinates of the polygon.
   ; py - The y coordinates of the polygon.
   ;
   ; The return value of the function is 1 if the point is inside the
   ; polygon and 0 if it is outside the polygon.

       sx = Size(px)
       sy = Size(py)
       IF (sx[0] EQ 1) THEN NX=sx[1] ELSE RETURN, -1    ; Error if px not a vector
       IF (sy[0] EQ 1) THEN NY=sy[1] ELSE RETURN, -1    ; Error if py not a vector
       IF (NX EQ NY) THEN N = NX ELSE RETURN, -1        ; Incompatible dimensions
       
       tmp_px = [px, px[0]]                             ; Close Polygon in x
       tmp_py = [py, py[0]]                             ; Close Polygon in y
        
       i = indgen(N)                                    ; Counter (0:NX-1)
       ip = indgen(N)+1                                 ; Counter (1:nx)
        
       X1 = tmp_px(i)  - x 
       Y1 = tmp_py(i)  - y
       X2 = tmp_px(ip) - x 
       Y2 = tmp_py(ip) - y
       
       dp = X1*X2 + Y1*Y2                               ; Dot-product
       cp = X1*Y2 - Y1*X2                               ; Cross-product
       theta = Atan(cp,dp)
       
       IF (Abs(Total(theta)) GT !PI) THEN RETURN, 1 ELSE RETURN, 0
   END

lib/Geo/Hash.pm
package Geo::Hash;

use warnings;
use strict;
use Carp;

=head1 NAME

Geo::Hash - Encode / decode geohash.org locations.

=head1 VERSION

This document describes Geo::Hash version 0.03

=cut

our $VERSION = '0.03';

=head1 SYNOPSIS

    use Geo::Hash;

    my $gh = Geo::Hash->new;
    my $hash = $gh->encode( $lat, $lon );
    my ( $lat, $lon ) = $gh->decode( $hash );
  
=head1 DESCRIPTION

Geohash is a latitude/longitude geocode system invented by Gustavo
Niemeyer when writing the web service at geohash.org, and put into the
public domain.

This module encodes and decodes geohash locations.

See L<http://en.wikipedia.org/wiki/Geohash> and L<http://geohash.org>
for more information.

=head1 INTERFACE 

=head2 C<< new >>

Create a new Geo::Hash object.

    my $gh = Geo::Hash->new;

=cut

sub new { bless {}, shift }

my @ENC = qw(
  0 1 2 3 4 5 6 7 8 9 b c d e f g h j k m n p q r s t u v w x y z
);

my %DEC = map { $ENC[$_] => $_ } 0 .. $#ENC;

sub _mid {
    my ( $ar, $wh ) = @_;
    return ( $ar->[$wh][0] + $ar->[$wh][1] ) / 2;
}

# The number of bits necessary to represent the specified number of
# decimal digits
sub _d2b { int( shift() * 3.32192809488736 + 1 ) }

sub _bits_for_number {
    my $n = shift;
    return 0 unless $n =~ s/.*\.//;
    return _d2b( length $n );
}

=head2 C<< precision >>

Infer a suitable precision (number of characters in hash) for a given
lat, lon pair.

    my $prec = $gh->precision( $lat, $lon );

=cut

sub precision {
    my ( $self, $lat, $lon ) = @_;
    my $lab = _bits_for_number( $lat ) + 8;
    my $lob = _bits_for_number( $lon ) + 9;
    return int( ( ( $lab > $lob ? $lab : $lob ) + 1 ) / 2.5 );
}

=head2 C<< encode >>

Encode a lat, long pair into a geohash.

    my $hash = $gh->encode( $lat, $lon );

You may optionally supply the length of the desired geohash:

    # Very precise
    my $hash = $gh->encode( $lat, $lon, 10 );

If the precision argument is omitted C<precision> will be used to
provide a default.

=cut

sub encode {
    croak "encode needs two or three arguments"
      unless @_ >= 3 && @_ <= 4;
    my ( $self, @pos ) = splice @_, 0, 3;
    my $prec = shift || $self->precision( @pos );
    my $int  = [ [ 90, -90 ], [ 180, -180 ] ];
    my $flip = 1;
    my @enc  = ();
    while ( @enc < $prec ) {
        my $bits = 0;
        for ( 0 .. 4 ) {
            my $mid = _mid( $int, $flip );
            my $bit = $pos[$flip] >= $mid ? 1 : 0;
            $bits = ( ( $bits << 1 ) | $bit );
            $int->[$flip][$bit] = $mid;
            $flip ^= 1;
        }
        push @enc, $ENC[$bits];
    }
    return join '', @enc;
}

=head2 C<< decode_to_interval >>

Like C<decode> but instead of returning a pair of coordinates returns
the interval for each coordinate. This gives some indication of how
precisely the original hash specified the location.

The return value is a pair of array refs. Each referred to array
contains the upper and lower bounds for each coordinate.

    my ( $lat_range, $lon_range ) = $gh->decode_to_interval( $hash );
    # $lat_range and $lon_range are references to two element arrays

=cut

sub decode_to_interval {
    croak "Needs one argument"
      unless @_ == 2;
    my ( $self, $hash ) = @_;

    my $int = [ [ 90, -90 ], [ 180, -180 ] ];
    my $flip = 1;

    for my $ch ( split //, $hash ) {
        if ( defined( my $bits = $DEC{$ch} ) ) {
            for ( 0 .. 4 ) {
                $int->[$flip][ ( $bits & 16 ) >> 4 ]
                  = _mid( $int, $flip );
                $flip ^= 1;
                $bits <<= 1;
            }
        }
        else {
            croak "Bad character '$ch' in hash '$hash'";
        }
    }

    return @$int;
}

=head2 C<< decode >>

Decode a geohash into a lat, long pair.

    my ( $lat, $lon ) = $gh->decode( $hash );

=cut

sub decode {
    my @int = shift->decode_to_interval( @_ );
    return map { _mid( \@int, $_ ) } 0 .. 1;
}

1;
__END__

=head1 CONFIGURATION AND ENVIRONMENT
  
Geo::Hash requires no configuration files or environment variables.

=head1 DEPENDENCIES

None.

=head1 INCOMPATIBILITIES

None reported.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-geo-hash@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.

=head1 AUTHOR

Andy Armstrong  C<< <andy@hexten.net> >>

L<http://geohash.org/gcwrdtsvrfgr>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2008, Andy Armstrong C<< <andy@hexten.net> >>.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


andy/geohash.c
/* geohash.c */

#include <string.h>
#include <stdio.h>
#include <math.h>

#define AVERAGE(d1, d2) (((d1) + (d2)) / 2)
#define BAD_CODE 0xFF

static const char *_enc = "0123456789bcdefghjkmnpqrstuvwxyz";
static unsigned char _dec[256];
static int _dec_inited = 0;

/* The buffer must be at least len + 1 bytes long */
char *
encode_with_len( double lat, double lon, char *hash, size_t len ) {
    double interval[2][2] = { {90, -90}, {180, -180} };
    double pos[2];
    int hp, flip = 1;
    pos[0] = lat;
    pos[1] = lon;

    for ( hp = 0; hp < len; hp++ ) {
        unsigned b, bits = 0;
        for ( b = 0; b < 5; b++ ) {
            double mid = AVERAGE( interval[flip][0], interval[flip][1] );
            unsigned bit = pos[flip] >= mid ? 1 : 0;
            bits = ( bits << 1 ) | bit;
            interval[flip][bit] = mid;
            flip ^= 1;
        }
        hash[hp] = _enc[bits];
    }

    hash[hp] = '\0';
    return hash;
}

unsigned
_bits_for_number( double n ) {
    char buf[40];
    int bp;
    const char *dot;
    sprintf( buf, "%32.8f", n );

    if ( ( dot = strchr( buf, '.' ) ) == NULL ) {
        return 0;
    }

    for ( bp = strlen( buf ) - 1; bp >= 0; bp-- ) {
        if ( buf[bp] != '0' ) {
            return ( int ) ( ( double ) ( bp - ( dot - buf ) ) *
                             3.32192809488736 + 1 );
        }
    }

    return 0;
}

int
precision( double lat, double lon ) {
    int lab = _bits_for_number( lab ) + 8;
    int lob = _bits_for_number( lon ) + 9;
    return ( int ) ( ( ( lab > lob ? lab : lob ) + 1 ) / 2.5 );
}

char *
encode( double lat, double lon, char *hash, size_t len ) {
    int prec = precision( lat, lon );
    if ( len > prec ) {
        len = prec;
    }
    return encode_with_len( lat, lon, hash, lon );
}

static void
_init_hash_to_bits( void ) {
    if ( !_dec_inited ) {
        int ep;
        memset( _dec, BAD_CODE, sizeof( _dec ) );
        for ( ep = 0; _enc[ep]; ep++ ) {
            _dec[( int ) _enc[ep]] = ep;
        }
        _dec_inited = 1;
    }
}

int
decode_to_interval( const char *hash, double *lat_lo, double *lon_lo,
                    double *lat_hi, double *lon_hi ) {
    double interval[2][2] = { {90, -90}, {180, -180} };
    unsigned flip = 1;
    const unsigned char *hp;

    _init_hash_to_bits(  );

    for ( hp = ( const unsigned char * ) hash; *hp; hp++ ) {
        unsigned b, bits = _dec[*hp];
        if ( bits == BAD_CODE ) {
            return -1;
        }
        for ( b = 0; b < 5; b++ ) {
            interval[flip][( bits & 0x10 ) >> 4] =
                AVERAGE( interval[flip][0], interval[flip][1] );
            flip ^= 1;
            bits <<= 1;
        }
    }

    *lat_lo = interval[0][1];
    *lat_hi = interval[0][0];
    *lon_lo = interval[1][1];
    *lon_hi = interval[1][0];

    return 0;
}

int
decode( const char *hash, double *lat, double *lon ) {
    double lat_lo, lon_lo, lat_hi, lon_hi;
    int rc;

    if ( ( rc =
           decode_to_interval( hash, &lat_lo, &lon_lo, &lat_hi,
                               &lon_hi ) ) != 0 ) {
        return rc;
    }

    *lat = AVERAGE( lat_lo, lat_hi );
    *lon = AVERAGE( lon_lo, lon_hi );

    return rc;
}

static int _test_num = 0;

void
ok( int pass, const char *msg ) {
    printf( "%sok %d %s\n", pass ? "" : "not ", ++_test_num, msg );
}

void
plan( int count ) {
    printf( "1..%d\n", count );
}

#define COUNT(ar) (sizeof(ar) / sizeof(ar[0]))

int
main( void ) {
    /* *INDENT-OFF* */
    static struct {
        const char *hash;
        double lat, lon;
        double eps;
    } tests[] = {
        { "ezs42",        42.6,       -5.6,    0.01     }, 
        { "mh7w",        -20,         50,      0.1      }, 
        { "t3b9m",        10.1,       57.2,    0.1      },
        { "c2b25ps",      49.26,    -123.26,   0.01     }, 
        { "80021bgm",      0.005,   -179.567,  0.001    },
        { "k484ht99h2",  -30.55555,    0.2,    0.00001  }, 
        { "8buh2w4pnt",    5.00001, -140.6,    0.00001  }
    };
    int tn;
    
    /* *INDENT-ON* */
    plan( COUNT( tests ) * 4 );

    for ( tn = 0; tn < COUNT( tests ); tn++ ) {
        char buf[100];
        double lat, lon;
        char *hash = encode_with_len( tests[tn].lat, tests[tn].lon, buf,
                                      strlen( tests[tn].hash ) );
        ok( strcmp( hash, tests[tn].hash ) == 0, "encode_with_len" );
        // fprintf( stderr, "# %s %s\n", tests[tn].hash, hash );
        ok( 0 == decode( hash, &lat, &lon ), "decode" );
        // fprintf( stderr, "# %g %g | %g %g\n", lat, tests[tn].lat, lon,
        //          tests[tn].lon );
        ok( fabs( lat - tests[tn].lat ) < tests[tn].eps, "lat" );
        ok( fabs( lon - tests[tn].lon ) < tests[tn].eps, "lon" );
    }

    return 0;
}
